js/modules/k6/execution/execution.go

```diff
@@ -0,0 +1,170 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package execution
+
+import (
+	"context"
+	"errors"
+
+	"github.com/dop251/goja"
+
+	"go.k6.io/k6/js/common"
+	"go.k6.io/k6/lib"
+)
+
+// Execution is a JS module to return information about the execution in progress.
+type Execution struct{}
+
+// New returns a pointer to a new Execution.
+func New() *Execution {
+	return &Execution{}
+}
+
+// GetVUStats returns information about the currently executing VU.
+func (e *Execution) GetVUStats(ctx context.Context) (goja.Value, error) {
+	vuState := lib.GetState(ctx)
+	if vuState == nil {
+		return nil, errors.New("getting VU information in the init context is not supported")
+	}
+
+	rt := common.GetRuntime(ctx)
+	if rt == nil {
+		return nil, errors.New("goja runtime is nil in context")
+	}
+
+	scID, _ := vuState.GetScenarioVUID()
+	stats := map[string]interface{}{
+		"id":         vuState.Vu,
+		"idScenario": scID,
+		"iteration":  vuState.GetIteration(),
+		"iterationScenario": func() goja.Value {
+			return rt.ToValue(vuState.GetScenarioVUIter())
+		},
+	}
+
+	obj, err := newLazyJSObject(rt, stats)
+	if err != nil {
+		return nil, err
+	}
+
+	return obj, nil
+}
+
+// GetScenarioStats returns information about the currently executing scenario.
+func (e *Execution) GetScenarioStats(ctx context.Context) (goja.Value, error) {
+	ss := lib.GetScenarioState(ctx)
+	vuState := lib.GetState(ctx)
+	if ss == nil || vuState == nil {
+		return nil, errors.New("getting scenario information in the init context is not supported")
+	}
+
+	rt := common.GetRuntime(ctx)
+	if rt == nil {
+		return nil, errors.New("goja runtime is nil in context")
+	}
+
+	var iterGlobal interface{}
+	// This stinks...
+	if ig := vuState.GetScenarioGlobalVUIter(); ig < 0 {
+		iterGlobal = goja.Null()
+	} else {
+		iterGlobal = ig
+	}
+
+	stats := map[string]interface{}{
+		"name":      ss.Name,
+		"executor":  ss.Executor,
+		"startTime": ss.StartTime,
+		"progress": func() goja.Value {
+			p, _ := ss.ProgressFn()
+			return rt.ToValue(p)
+		},
+		"iteration":       vuState.GetScenarioLocalVUIter(),
+		"iterationGlobal": iterGlobal,
+	}
+
+	obj, err := newLazyJSObject(rt, stats)
+	if err != nil {
+		return nil, err
+	}
+
+	return obj, nil
+}
+
+// GetTestStats returns global test information.
+func (e *Execution) GetTestStats(ctx context.Context) (goja.Value, error) {
+	es := lib.GetExecutionState(ctx)
+	if es == nil {
+		return nil, errors.New("getting test information in the init context is not supported")
+	}
+
+	rt := common.GetRuntime(ctx)
+	if rt == nil {
+		return nil, errors.New("goja runtime is nil in context")
+	}
+
+	stats := map[string]interface{}{
+		// XXX: For consistency, should this be startTime instead, or startTime
+		// in ScenarioStats be converted to duration?
+		"duration": func() goja.Value {
+			return rt.ToValue(es.GetCurrentTestRunDuration().String())
+		},
```

Comment:
I think this might be better off as `progress` although I also would expect people would want the max duration of the test which might be put here as `duration` ;)

Reply:
Hhmm so what's your suggestion? :smile:

Rename this to `progress` and convert it to float, and add a (max)`duration` property?

Reply:
I think `duration` is fine (and we might add `maxDuration` in the future, not now). Though I'd like it to be a float number in milliseconds instead of a string, so JS people don't have to parse stringy times.

Reply:
Resolved by 7f420594.